runtime/compiler/optimizer/JProfilingValue.cpp

```diff
@@ -193,104 +193,129 @@ TR_JProfilingValue::perform()
                performTransformation(comp(), "%s Removing profiling treetop, node n%dn is already profiled\n",
                   optDetailString(), value->getGlobalIndex()))
             /* Found that second and third child were having more than one ref counts. 
             TR_ASSERT_FATAL(node->getFirstChild()->getSecondChild()->getReferenceCount() == 1 &&
                node->getFirstChild()->getThirdChild()->getReferenceCount() == 1,
                "Second and Third Child of the value calls should be referenced only once");
             */
+            {
             TR::TransformUtil::removeTree(comp(), cursor);
+            }
          else
+            {
             alreadyProfiledValues->set(value->getGlobalIndex());
+            }
          
          }
       // Emptying a bit vector after scanning whole extended basic blocks will keep number of bits set in bit vector low.
-      else if (cursor->getNode()->getOpCodeValue() == TR::BBStart && !cursor->getNode()->getBlock()->isExtensionOfPreviousBlock())
+      else if (node->getOpCodeValue() == TR::BBStart && !node->getBlock()->isExtensionOfPreviousBlock())
          {
          alreadyProfiledValues->empty();
          }
-      cursor = cursor->getNextTreeTop();
-      }    
+      else
+         {
+         performOnNode(node, cursor, alreadyProfiledValues, &checklist);
+         }
+      cursor = nextTT;
+      }
    }
+
 void
-TR_JProfilingValue::performOnNode(TR::Node *node, TR::TreeTop *tt, TR::NodeChecklist *checklist)
+TR_JProfilingValue::performOnNode(TR::Node *node, TR::TreeTop *cursor, TR_BitVector *alreadyProfiledValues, TR::NodeChecklist *checklist)
    {
    if (checklist->contains(node))
       return;
    checklist->add(node);
 
-   if (node->getOpCode().isCall() && !node->getByteCodeInfo().doNotProfile() &&
-       (node->getSymbol()->getMethodSymbol()->isVirtual() ||
-        node->getSymbol()->getMethodSymbol()->isInterface()))
-      addProfiling(node->getFirstChild(), tt);
+   TR::TreeTop *preceedingTT = NULL;
+   TR::Node *profiledNode = NULL;
+   TR::Node *constNode = NULL;
+   TR::SymbolReference *profiler = NULL;
+   if (node->getOpCode().isCallIndirect() && !node->getByteCodeInfo().doNotProfile() &&
+         (node->getSymbol()->getMethodSymbol()->isVirtual() || 
+            node->getSymbol()->getMethodSymbol()->isInterface()))
+      {
+      /**
+       * RecompilationModifier adds the profiling placeholder call for Virtual calls. 
+       * Quickly check the next treetop, if it is profiling same virtual call target. In that case we do not need to add
+       * Profiling Call as it is already there.
+       */
+      profiledNode = node->getFirstChild();
+      TR::Node *nextTTNode = cursor->getNextTreeTop() ? cursor->getNextTreeTop()->getNode() : NULL;
+      if (!(alreadyProfiledValues->isSet(profiledNode->getGlobalIndex()) || (nextTTNode && nextTTNode->isProfilingCode() &&
```

Comment:
can you format this to be a bit more readable? I'd suggest dropping the || to be the start of a newline and having all the &&s at the start of a line rather than the end with a much lower indent... I've found this hard to read while reading the code...

Reply:
I'm a bit confused what the semantic meaning of the case after the || - could you explain?

Reply:
I have this quick check to see if the next Tree Top is profiling the same VFT load of receiver object or not. For Virtual calls, we profile VFT Load of receiver object and in other opts we add placeholder call to profile it after the virtual call. This is a quick check to see if value is already being profiled in next tree then do not add the placeholder call. This is to avoid adding un-necessary trees which eventually will be cleaned up as we are scanning for duplicate profiling candidates. Let me restructure if condition and comment this properly.

Reply:
I have re-structure the if condition and moved comment closer to if statement in https://github.com/eclipse/openj9/pull/6585/commits/21e100bcbb8f52db7d0e1ac3fc4af3ba1bdf0892